### PR TITLE
Parser allow specify primary key in column list

### PR DESCRIPTION
**Pull request by kitaisreal**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

https://github.com/ClickHouse/ClickHouse/issues/15591

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow specify primary key in column list of CREATE TABLE query.

### CODE REVIEW

File: src/Parsers/ParserCreateQuery.cpp [hunk 1]
```diff
@@ -210,6 +223,14 @@ bool ParserTablePropertiesDeclarationList::parseImpl(Pos & pos, ASTPtr & node, E
             indices->children.push_back(elem);
         else if (elem->as<ASTConstraintDeclaration>())
             constraints->children.push_back(elem);
+        else if (elem->as<ASTIdentifier>() || elem->as<ASTFunction>())
+        {
+            if (primary_key)
+            {
+                throw Exception("Multiple primary keys are not allowed.", ErrorCodes::LOGICAL_ERROR);
```
Comment (alesapin) [inline src/Parsers/ParserCreateQuery.cpp:230]: `LOGICAL_ERROR` identifies error unrecoverable by the user. I think `BAD_ARGUMENTS` will be better.

File: src/Parsers/ParserCreateQuery.cpp [hunk 2]
```diff
@@ -467,6 +490,15 @@ bool ParserCreateTableQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expe
     query->set(query->columns_list, columns_list);
     query->set(query->storage, storage);
 
+    if (query->storage && query->columns_list && query->columns_list->primary_key)
+    {
+        if (query->storage->primary_key)
+        {
+            throw Exception("Multiple primary keys are not allowed.", ErrorCodes::LOGICAL_ERROR);
```
Comment (alesapin) [inline src/Parsers/ParserCreateQuery.cpp:497]: Same here for an exception.

**Review by alesapin — CHANGES_REQUESTED** [review]
Hi!

Thank you for your contribution. The code looks almost OK, but I think we should add functional tests for the new feature. You can find a lot of examples in `tests/queries/0_stateless` directory.

File: src/Parsers/ParserCreateQuery.cpp (hunk 2 above)
Comment (kitaisreal) [reply src/Parsers/ParserCreateQuery.cpp]: Fixed. Thanks.

File: src/Parsers/ParserCreateQuery.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Parsers/ParserCreateQuery.cpp]: Fixed. Thanks.

**Comment by kitaisreal** [comment]
@alesapin Thanks for review. Added functional test.

**Comment by alesapin** [comment]
> @alesapin Thanks for review. Added functional test.

Ok, but maybe we can also check that our `PRIMARY KEY` works? For example, create `ReplacingMergeTree` with some primary key, insert several rows with equal primary key value and execute `SELECT .. FINAL`. If everything works well we should see only one row. Also I recommend adding `DETACH TABLE ...` and `ATTACH TABLE ...` queries just to be sure that parser works well.

**Comment by kitaisreal** [comment]
@alesapin agreed. Updated tests with ATTACH DETACH and ReplacingMergeTree engine.

**Comment by kitaisreal** [comment]
@alesapin could you please help with AST fuzzer check. Or how can I reproduce this locally ? Thanks.

**Comment by alesapin** [comment]
> @alesapin could you please help with AST fuzzer check. Or how can I reproduce this locally ? Thanks.

Just don't care about it. All failures are known.

**Comment by alesapin** [comment]
Can you add one more case, when we have `PRIMARY KEY` in new syntax, but no `ORDER BY`? Also, it would be nice to check some complex expression, like `PRIMARY KEY(x, toString(x * y))`.

**Comment by alesapin** [comment]
Ok, I'll add several tests by myself.

**Comment by kitaisreal** [comment]
@alesapin Okay, thanks.

**Review by alesapin — APPROVED** [review]

**Pull request merged**